File: TutorialSamples/ControlSamplesWpf/ZoomPanControlSampleVB/ZoomPanDimensionConverter.vb
Imports System
Imports System.Globalization
Imports System.Windows.Markup
Imports SharpVectors.Runtime

<MarkupExtensionReturnType(GetType(Double))>
Public NotInheritable Class ZoomPanDimensionConverter
    Inherits MarkupExtension
    Implements IMultiValueConverter

    Public Overrides Function ProvideValue(ByVal serviceProvider As IServiceProvider) As Object
        Return Me
    End Function

    Public Function Convert(ByVal values As Object(), ByVal targetType As Type,
                            ByVal parameter As Object, ByVal culture As CultureInfo) _
                            As Object Implements IMultiValueConverter.Convert
        Dim zoomPanControl = TryCast(values(3), ZoomPanControl)
        If values(0) Is Nothing OrElse zoomPanControl Is Nothing Then
            Return DependencyProperty.UnsetValue
        End If
        Dim size = CDbl(values(0))
        Dim offset = CDbl(values(1))
        Dim zoom = CDbl(values(2))
        Dim isWidth = String.Equals(parameter?.ToString(), "width", StringComparison.OrdinalIgnoreCase)
        Return Math.Max(If(isWidth, Math.Min(zoomPanControl.ExtentWidth / zoom - offset, size),
                        Math.Min(zoomPanControl.ExtentHeight / zoom - offset, size)), 0)
    End Function

    Public Function ConvertBack(ByVal value As Object, ByVal targetTypes As Type(),
                                ByVal parameter As Object, ByVal culture As CultureInfo) _
                                As Object() Implements IMultiValueConverter.ConvertBack
        Throw New NotImplementedException()
    End Function
End Class

File: Samples/WpfTestSvgControl/ZoomPanDimensionConverter.cs
using System;
using System.Globalization;

using System.Windows;
using System.Windows.Data;
using System.Windows.Markup;

using SharpVectors.Runtime;

namespace WpfTestSvgControl
{
    [MarkupExtensionReturnType(typeof(double))]
    public sealed class ZoomPanDimensionConverter : MarkupExtension, IMultiValueConverter
    {
        public override object ProvideValue(IServiceProvider serviceProvider)
        {
            return this;
        }

        public object Convert(object[] values, Type targetType, object parameter, CultureInfo culture)
        {
            //NOTE: Cannot pass ExtentWidth or ExtentHeight as one of the values because it does not seem to update
            var zoomPanControl = values[3] as ZoomPanControl;
            if (values[0] == null || zoomPanControl == null)
                return DependencyProperty.UnsetValue;
            var size   = (double)values[0];
            var offset = (double)values[1];
            var zoom   = (double)values[2];

            var isWidth = string.Equals(parameter?.ToString(), "width", StringComparison.OrdinalIgnoreCase);
            return Math.Max(isWidth ? Math.Min(zoomPanControl.ExtentWidth / zoom - offset, size)
                 : Math.Min(zoomPanControl.ExtentHeight / zoom - offset, size), 0);
        }

        public object[] ConvertBack(object value, Type[] targetTypes, object parameter, CultureInfo culture)
        {
            throw new NotImplementedException();
        }
    }

}
